### PR TITLE
Refactor TimestampConversion to return velox::Expected

### PR DESCRIPTION
**Pull request by NEUpanning**
We have recently introduced velox::Expected (https://github.com/facebookincubator/velox/pull/9858) to use in non-void no-throw APIs. It's more proper to return that.

### CODE REVIEW

**Comment by NEUpanning** [comment]
@rui-mo @PHILO-HE Could you help to review this PR please? Thanks.

**Review by PHILO-HE — COMMENTED** [review]
Looks good to me.
One minor suggestion: use `auto expected` for statements like `Expected<int64_t> expected = xxx;` to reduce redundancy.

**Review by rui-mo — COMMENTED** [review]
Thanks. Could you share some background for this change in the PR description?

**Comment by NEUpanning** [comment]
@rui-mo Thanks. Updated.

**Comment by NEUpanning** [comment]
@PHILO-HE Thanks for your suggestion. In my opinion, we need to know the return type to declare `daysSinceEpoch`.
```
Expected<int64_t> expected = util::daysSinceEpochFromDate(year, month, day);
   ...
int64_t daysSinceEpoch = expected.value();
```

**Review by PHILO-HE — COMMENTED** [review]
I have no other comment. Thanks!

File: velox/type/TimestampConversion.cpp [hunk 1]
```diff
@@ -158,10 +158,11 @@ bool isValidWeekOfMonthDate(
   }
 
   int64_t daysSinceEpochOfFirstDayOfMonth;
-  const Status status =
-      daysSinceEpochFromDate(year, month, 1, daysSinceEpochOfFirstDayOfMonth);
-  if (!status.ok()) {
+  Expected<int64_t> expected = daysSinceEpochFromDate(year, month, 1);
+  if (expected.hasError()) {
     return false;
+  } else {
```
Comment (rui-mo) [inline velox/type/TimestampConversion.cpp:164]: Perhaps remove the 'else' after 'return'. Ditto for the others.

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: velox/type/TimestampConversion.cpp (hunk 1 above)
Comment (NEUpanning) [reply velox/type/TimestampConversion.cpp]: Thanks. Updated.

File: velox/type/TimestampConversion.cpp [hunk 2]
```diff
@@ -586,26 +598,27 @@ bool isValidDayOfYear(int32_t year, int32_t dayOfYear) {
...
-daysSinceEpochFromDate(int32_t year, int32_t month, int32_t day, int64_t& out) {
+Expected<int64_t>
+daysSinceEpochFromDate(int32_t year, int32_t month, int32_t day) {
   int64_t daysSinceEpoch = 0;
 
   if (!isValidDate(year, month, day)) {
     if (threadSkipErrorDetails()) {
-      return Status::UserError();
+      return folly::makeUnexpected(Status::UserError());
     } else {
```
Comment (rui-mo) [inline velox/type/TimestampConversion.cpp:619]: nit: remove else

File: velox/type/TimestampConversion.cpp [hunk 3]
```diff
@@ -620,29 +633,30 @@ daysSinceEpochFromDate(int32_t year, int32_t month, int32_t day, int64_t& out) {
...
     int32_t weekOfYear,
-    int32_t dayOfWeek,
-    int64_t& out) {
+    int32_t dayOfWeek) {
   if (!isValidWeekDate(weekYear, weekOfYear, dayOfWeek)) {
-    return Status::UserError(
-        "Date out of range: {}-{}-{}", weekYear, weekOfYear, dayOfWeek);
+    if (threadSkipErrorDetails()) {
+      return folly::makeUnexpected(Status::UserError());
+    } else {
```
Comment (rui-mo) [inline velox/type/TimestampConversion.cpp:646]: ditto

File: velox/type/TimestampConversion.cpp [hunk 4]
```diff
@@ -677,35 +691,37 @@ Expected<int64_t> daysSinceEpochFromWeekOfMonthDate(
...
 }
 
-Status
-daysSinceEpochFromDayOfYear(int32_t year, int32_t dayOfYear, int64_t& out) {
+Expected<int64_t> daysSinceEpochFromDayOfYear(int32_t year, int32_t dayOfYear) {
   if (!isValidDayOfYear(year, dayOfYear)) {
-    return Status::UserError("Day of year out of range: {}", dayOfYear);
+    if (threadSkipErrorDetails()) {
+      return folly::makeUnexpected(Status::UserError());
+    } else {
```
Comment (rui-mo) [inline velox/type/TimestampConversion.cpp:716]: ditto

**Review by rui-mo — APPROVED** [review]
Thanks. Looks good to me overall.

**Comment by NEUpanning** [comment]
@mbasmanova Could you help to review this PR please? Thanks.

File: velox/functions/lib/DateTimeFormatter.cpp [hunk 5]
```diff
@@ -1423,27 +1423,25 @@ Expected<DateTimeResult> DateTimeFormatter::parse(
   }
 
   // Convert the parsed date/time into a timestamp.
-  int64_t daysSinceEpoch;
-  Status status;
+  Expected<int64_t> expected;
```
Comment (mbasmanova) [inline velox/functions/lib/DateTimeFormatter.cpp:1426]: 'expected' is not very descriptive name; why not 'daysSinceEpoch'?

File: velox/functions/prestosql/DateTimeFunctions.h [hunk 6]
```diff
@@ -402,42 +402,37 @@ struct LastDayOfMonthFunction : public InitSessionTimezone<T>,
       const arg_type<Timestamp>& timestamp) {
     auto dt = getDateTime(timestamp, this->timeZone_);
     int64_t daysSinceEpochFromDate;
-    auto status =
-        util::lastDayOfMonthSinceEpochFromDate(dt, daysSinceEpochFromDate);
-    if (!status.ok()) {
-      VELOX_DCHECK(status.isUserError());
-      VELOX_USER_FAIL(status.message());
+    Expected<int64_t> expected = util::lastDayOfMonthSinceEpochFromDate(dt);
```
Comment (mbasmanova) [inline velox/functions/prestosql/DateTimeFunctions.h:405]: expected -> daysSinceEpochFromDate

**Review by mbasmanova — APPROVED** [review]
@NEUpanning Nice refactoring, but using 'expected' for variable names is not very readable. Why can't we use the original names?

**Comment by NEUpanning** [comment]
@mbasmanova Thanks! I've replaced variable names.

**Pull request closed without merging**